schema205/header_entries.py

```diff
@@ -0,0 +1,637 @@
+    def __str__(self):
+        s = ''
+        for p in self._preamble:
+            s += p
+        s += '\n'
+        s += self._doxynotes
+        s += '\n'
+        if self._is_top_container:
+            s += self._namespace.value
+        else:
+            s += self._top_namespace.value
+        s += '\n'
+        for e in self._epilogue:
+            s += e
+        return s
+
+    @property
+    def root(self):
+        if self._is_top_container:
+            return self._namespace
+        else:
+            return self._top_namespace
+
+    # .............................................................................................
+    @staticmethod
+    def modified_insertion_sort(obj_list):
+        '''From https://stackabuse.com/sorting-algorithms-in-python/#insertionsort'''
+        swapped = False
+        # Start on the second element as we assume the first element is sorted
+        for i in range(1, len(obj_list)):
+            item_to_insert = obj_list[i]
+            # And keep a reference of the index of the previous element
+            j = i - 1
+            # Move all items of the sorted segment forward if they are larger than the item to insert
+            while j >= 0 and any(obj > item_to_insert for obj in obj_list[0:j+1]):
+                obj_list[j + 1] = obj_list[j]
+                swapped = True
+                j -= 1
+            # Insert the item
+            obj_list[j + 1] = item_to_insert
+        return swapped
+
+    # .............................................................................................
+    def translate(self, input_file_path, container_class_name, schema_base_class_name=None):
+        '''X'''
+        self._source_dir = os.path.dirname(os.path.abspath(input_file_path))
+        self._schema_name = os.path.splitext(os.path.splitext(os.path.basename(input_file_path))[0])[0]
+        self._references.clear()
+        self._fundamental_data_types.clear()
+        self._preamble.clear()
+        self._epilogue.clear()
+
+        self._contents = load(input_file_path)
+
+        # If container_class_name is empty, I must be the top container. Also true if container_class_name is my name.
+        self._is_top_container = not container_class_name or (container_class_name == self._schema_name)
+
+        self._fundamental_base_class = schema_base_class_name if schema_base_class_name else ''
+
+        # Load meta info first (assuming that base level tag == Schema means object type == Meta)
+        self._load_meta_info(self._contents['Schema'])
+        self._add_include_guard(self._schema_name)
+        self._add_included_headers(self._contents['Schema'].get('References'))
+
+        # Create "root" node(s)
+        if self._is_top_container:
+            self._namespace = Header_entry(f'{self._schema_name}_NS')
+        else:
+            self._top_namespace = Header_entry(f'{container_class_name}_NS')
```

Comment:
I think we need to figure out the difference between the schema namespace and the container namespace. We have an unfortunate situation where one of our schema files shares the name with the overall schema name. I think these need to be differentiated better. For example, the namespace in ASHRAE205.h should be named after its schema file (ASHRAE205_NS), but it needs to be wrapped by the overall schema namespace which shouldn't be the same name. I think this is the intention of what we see in the RSXXXX.h files where RSXXXX_NS is wrapped in ASHRAE205_NS.

Reply:
Agreed. When we had RS "is-a" ASHRAE205, it got convoluted, and we should sort how to de-convolute it.